libs/dyn/pattern_trie.go

```diff
@@ -0,0 +1,200 @@
+		}
+
+		if next == nil {
+			return fmt.Errorf("invalid component type: %T", component)
+		}
+
+		// Mark as end of pattern if this is the last component.
+		if !next.isEnd && i == len(pattern)-1 {
+			next.isEnd = true
+		}
+
+		// Move to next node
+		current = next
+	}
+
+	return nil
+}
+
+// SearchPath checks if the given path matches any pattern in the trie.
+// A path matches if it exactly matches a pattern or if it matches a pattern with wildcards.
+func (t *PatternTrie) SearchPath(path Path) (Pattern, bool) {
+	// We statically allocate the prefix array that is used to track the current
+	// prefix accumulated while walking the prefix tree. Having the static allocation
+	// ensures that we do not allocate memory on every recursive call.
+	prefix := make(Pattern, len(path))
+	pattern, ok := t.searchPathRecursive(t.root, path, prefix, 0)
+	return pattern, ok
+}
+
+// searchPathRecursive is a helper function that recursively checks if a path matches any pattern.
+// Arguments:
+// - node: the current node in the trie.
+// - path: the path to check.
+// - prefix: the prefix accumulated while walking the prefix tree.
+// - index: the current index in the path / prefix
+//
+// Note we always expect the path and prefix to be the same length because wildcards like * and [*]
+// only match a single path component.
+func (t *PatternTrie) searchPathRecursive(node *trieNode, path Path, prefix Pattern, index int) (Pattern, bool) {
+	if node == nil {
+		return nil, false
+	}
+
+	// Zero case, when the query path is the root node. We return nil here to match
+	// the semantics of [MustPatternFromString] which returns nil for the empty string.
+	if len(path) == 0 {
+		return nil, node.isEnd
+	}
+
+	// If we've reached the end of the path, check if this node is a valid end of a pattern.
+	isLast := index == len(path)
+	if isLast {
+		return prefix, node.isEnd
+	}
+
+	currentComponent := path[index]
+
+	// First check if the key wildcard is set for the current index.
+	if currentComponent.isKey() && node.anyKey != nil {
+		prefix[index] = AnyKey()
+		pattern, ok := t.searchPathRecursive(node.anyKey, path, prefix, index+1)
+		if ok {
+			return pattern, true
+		}
+	}
+
+	// If no key wildcard is set, check if the key is an exact match.
+	if currentComponent.isKey() {
+		child, exists := node.pathKey[currentComponent.Key()]
+		if !exists {
+			return prefix, false
+		}
+		prefix[index] = currentComponent
+		return t.searchPathRecursive(child, path, prefix, index+1)
+	}
+
+	// First check if the index wildcard is set for the current index.
+	if currentComponent.isIndex() && node.anyIndex != nil {
+		prefix[index] = AnyIndex()
+		pattern, ok := t.searchPathRecursive(node.anyIndex, path, prefix, index+1)
+		if ok {
+			return pattern, true
+		}
+	}
+
+	// If no index wildcard is set, check if the index is an exact match.
+	if currentComponent.isIndex() {
+		child, exists := node.pathIndex[currentComponent.Index()]
+		if !exists {
+			return prefix, false
```

Comment:
are we supposed to return partial pattern? Is there a use case for that?

Reply:
Does not matter, the bool should take precedence when using this method. Modified the function to return nil here.